core/collection-factory/api/src/main/java/org/eclipse/rdf4j/collection/factory/api/CollectionFactory.java

```diff
@@ -0,0 +1,74 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.collection.factory.api;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+import org.eclipse.rdf4j.common.exception.RDF4JException;
+import org.eclipse.rdf4j.model.Value;
+
+/**
+ * A Factory that may generate optimised and/or disk based collections
+ *
+ * Factories like this should not be cached but created a new everytime as the closing is important if they are disk
+ * based.
```

Comment:
Can we add  a bit more context on the intended use? I'd like to make it clear if this is intended to be something only used internally or not.

Reply:
I am not sure yet if using this by end-users is a good idea or not. I should probably add `@experimental` and `@internaluseonly` for now.